USB HID: add product name and serial to descriptor.

enrikb opened the pull request:
I tried to add product name and serial number support to the internal descriptor for USB HID devices. If there is something similar for NFC devices it might be possible to add it, as well.

Tested in following environments:
 - Linux (Ubuntu 20.10, python3): OK (product name contains manufacturer)
 - Windows (Windows 10, python3): OK
 - MacOSX (Catalina, python3): OK
 - OpenBSD (6.8, python3): OK (Note:there have been changes in `struct usb_device_info`, see [OpenBSD bug mailing list](https://marc.info/?t=161565009600001&r=1&w=2). I don't know how this can be addressed to support older versions, too.)
 - FreeBSD (12.2, python3): OK (serial number support needs updated uhid-freebsd, see grembo/uhid-freebsd#1; fragmented CTAP messages did not work in my VM setup, but seems unrelated)

dainnilsson commented on examples/get_info.py:
```diff
@@ -51,6 +51,8 @@ def enumerate_devices():
 
 for dev in enumerate_devices():
     print("CONNECT: %s" % dev)
+    print("Product  name: %s" % dev.product_name)
```
Remove extra space.

enrikb replied:
Done.

dainnilsson reviewed (changes requested):
Thank you, this looks very promising!

In general the approach looks fine. One change I'd request is to use `None` when no value is present instead of an empty string.

For NFC, I suppose we could use the PCSC reader name instead of product name. I'm not sure how useful it is, but then again, maybe it's better than nothing?

I ran some quick tests of this on Ubuntu, seems to work fine there.
I also tried running this on Windows and it immediately broke things. It looks like it just needs to handle the case where there is no value present by returning `None` instead of raising a WinError.

I'll try to do some testing on the other OSes as well!

enrikb commented on fido2/hid/macos.py:
```diff
@@ -331,6 +342,30 @@ def get_int_property(dev, key):
+        return ""
+
+    try:
+        return out.raw.decode("utf-8")
```
Yes, it does.
Fixed.

enrikb commented on fido2/hid/macos.py:
```diff
@@ -331,6 +342,30 @@ def get_int_property(dev, key):
+
+    try:
+        return out.raw.decode("utf-8")
+    except Exception:
```
Changed `except` accordingly.

enrikb commented on fido2/hid/windows.py:
```diff
@@ -236,6 +238,28 @@ def get_vid_pid(device):
     return attributes.VendorID, attributes.ProductID
 
 
+def get_product_name(device):
```
`None` will be returned now.

enrikb commented:
I changed the code to return `None` whenever there is no serial number / product name or the empty string is returned by the respective native interface. Many interfaces do not seem to distinguish here. As an empty serial number or product name is quite useless anyway, `None` should be OK in either case as well as in most error cases.

Thank you very much for your feedback and the valuable hints for improvement!

dainnilsson commented:
Great, functionally looks good. One small improvement is to replace the pattern:
```
foobar = <expr>
foobar = foobar if foobar else None
```
With:
```
foobar = <expr> or None
```

I'm not done testing this on the different platforms yet, but so far I've run into some Python2 specific problems on Linux. I've also found a few other Python2 compatibility issues that are unrelated to this PR (like the use of `enum.auto`, I've pushed a fix for that on master). The plan is to drop support for Python 2 for the 1.0 release, but for now we do have a requirement to be Python 2 compatible.

`array.tobytes` doesn't exist. I think the _least ugly_ workaround is to replace:

`buf.tobytes().decode()` with `bytearray(buf).decode("utf-8")`.

I'll get back to you when I've done more testing with anything I find!

enrikb commented:
I rebased to include your Python 2 fixes and incorporated your latest feedback.
Unfortunately, it's getting harder to test Python 2 with current operating systems. But at least on MacOS (Catalina), `get_info.py` works fine using the system's default Python 2.

dainnilsson commented:
Everything looks good to me! I've run some very basic tests of the latest code on Ubuntu (py2 and 3), MacOS (x86, py2 and 3), FreeBSD (py 2 and 3) and OpenBSD (py3 only, couldn't get python 2 to run properly) and Windows 10 (py2 and 3). No issues found.

I'm not super happy about the FreeBSD code relying on an unmerged PR to uhid-freebsd (which looks like it hasn't been updated in some time), but I'm OK with making an exception here since it's so minor and shouldn't break anything.

Again, thanks for your work on this!

enrikb commented:
Thanks!
